src/osd/OSD.cc

```diff
@@ -9128,52 +8803,57 @@ void OSD::dequeue_op(
 
 
 void OSD::dequeue_peering_evt(
+  OSDShard *sdata,
   PG *pg,
   PGPeeringEventRef evt,
   ThreadPool::TPHandle& handle)
 {
-  auto curmap = service.get_osdmap();
   PG::RecoveryCtx rctx = create_context();
-  if (curmap->get_epoch() > pg->get_osdmap()->get_epoch()) {
-    advance_pg(curmap->get_epoch(), pg, handle, &rctx);
-  }
-  pg->do_peering_event(evt, &rctx);
-  auto need_up_thru = pg->get_need_up_thru();
-  auto same_interval_since = pg->get_same_interval_since();
-  dispatch_context_transaction(rctx, pg, &handle);
-  bool deleted = pg->is_deleted();
-  pg->unlock();
-
-  if (deleted) {
-    RWLock::WLocker l(pg_map_lock);
-    auto p = pg_map.find(pg->get_pgid());
-    if (p != pg_map.end() &&
-	p->second == pg) {
-      dout(20) << __func__ << " removed pg " << pg << " from pg_map" << dendl;
-      pg_map.erase(p);
-      pg->put("PGMap");
+  auto curmap = sdata->get_osdmap();
+  epoch_t need_up_thru = 0, same_interval_since = 0;
+  if (!pg) {
+    if (const MQuery *q = dynamic_cast<const MQuery*>(evt->evt.get())) {
+      handle_pg_query_nopg(*q);
     } else {
-      dout(20) << __func__ << " failed to remove pg " << pg << " from pg_map" << dendl;
+      derr << __func__ << " unrecognized pg-less event " << evt->get_desc() << dendl;
+      ceph_abort();
+    }
+  } else {
+    if (curmap->get_epoch() > pg->get_osdmap_epoch()) {
+      advance_pg(curmap->get_epoch(), pg, handle, &rctx);
+    }
+    pg->do_peering_event(evt, &rctx);
+    if (pg->is_deleted()) {
+      // do not dispatch rctx; the final _delete_some already did it.
+      discard_context(rctx);
```

Comment:
I liked this better than making discard_context a public interface of OSDService and having the PG call up.

Reply:
Well, do_peering_event() can return is_deleted(), and then the above code changes to
```
if (!pg->do_peering_event(evt, &rctx)) {
   // do not dispatch rctx; the PG was deleted
   discard_context(rctx);
```
That seems less clear to me, but the comment can make up for it.

I don't see a substantially better approach here unless we want to make a ton of changes to pass rctx all over the place, and reset it to be empty so that the final dispatch_context() is a no-op. PG deletion is kind of special.

Reply:
```OSD::discard_context()``` is just manipulating the ```PG::RecoveryCtx```, so you could move the implementation down to being a member function and invoke it cleanly from anybody who has one?

Reply:
Oh, good idea.  See the top 2 patches.  It's a bit of code churn but avoids this special case.

Reply:
...and that blows up completely because RecoveryCtx's lifecycle semantics are so fucking weird.  Every time I have to touch it I hate it more.  Can we please leave this cleanup for another day?

Reply:
Okay, I found the issue and pushed an alternate version.

Still hate RecoveryCtx.